Ensure that a module within a namespace package can be found by --pyargs

taschini opened the pull request:
This PR supersedes #1568 and addresses the problem highlighted in #1567 and in part #478.

See #1568 for the description of the test setup.

Unlike #1568, this PR does not attempt to emulate the import machinery to determine the filename of a  package or directory, delegating instead most of the work to the ``pkgutil`` module of the standard library.

**NB** In case of a non-top-level module or package, ``pkgutil.find_loader`` *will* import the parent package or packages.

nicoddemus commented on _pytest/main.py:
```diff
@@ -649,36 +645,29 @@ def _recurse(self, path):
+            return x
+        try:
+            path = loader.get_filename()
+        except:
```
It is usually better to avoid bare `except` clauses like this... could this be replaced by `except ImportError`?

nicoddemus commented on _pytest/main.py:
```diff
@@ -649,36 +645,29 @@ def _recurse(self, path):
+        try:
+            path = loader.get_filename()
+        except:
+            path = loader.modules[x][0].co_filename
```
In which situations is this necessary? A think a comment here would help. 

I tried to play a bit with this to see what it supposed to do, but didn't obtain much success:

```
Python 3.5.0 (v3.5.0:374f501f4567, Sep 13 2015, 02:16:59) [MSC v.1900 32 bit (Intel)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import pkgutil
>>> loader = pkgutil.find_loader('py')
>>> loader.modules['py']
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'SourceFileLoader' object has no attribute 'modules'
```

nicoddemus commented on testing/acceptance_test.py:
```diff
@@ -514,11 +515,20 @@ def test_pyargs_importerror(self, testdir, monkeypatch):
-            "*collected 0 items*"
-        ])
+
+        # Depending on whether the process running the test is the
```
You can force `testdir` to call pytest as a subprocess by calling `testdir.runpytest_subprocess`... seems better than try to accommodate both possibilities for `testdir.runpytest`.

nicoddemus commented on testing/acceptance_test.py:
```diff
@@ -557,6 +567,68 @@ def join_pythonpath(what):
+            cur = py.std.os.environ.get('PYTHONPATH')
+            if cur:
+                dirs += (cur,)
+            return ':'.join(str(p) for p in dirs)
```
Instead of `:` I think this should be `os.pathsep` (which is `;` on Windows)

taschini commented:
Appveyor caught a little problem with a test setup using PyPy on Windows: I'll fix that.

@nicoddemus I'll incorporate your feedback.  The reason why that try-except clause is needed is that sometimes ``_parsearg()`` gets invoked with ``AssertionRewritingHook``, which does not define a get_filename method, already in place.  Hence, the exception you'll get is an ``AttributeError``.

I'll catch explicitly that Exception and add a comment to explain why that is necessary.

taschini commented:
I incorporated the feedback from @nicoddemus and also fixed a problem caused in a test on Windows by a file left open by PyPy and not immediately garbage collected.

nicoddemus commented:
Thanks! :grin: 

LGTM. Could someone else take a look as well?

Also, we need a `CHANGELOG` entry and add @taschini to `AUTHORS` before merging.

taschini commented:
Shall I add the entries in ``CHANGELOG`` and ``AUTHORS`` or is it something that typically you guys do?

nicoddemus commented:
Please do. We usually ask for PR submitters to do it, but this has gone back and forth a few times already and I didn't want to ask for one more, hehehe. :sweat_smile:

nicoddemus commented:
It seems this is now in conflict. Would you please merge/rebase to fix it? Thanks!

taschini commented:
Rebased the PR.

nicoddemus commented:
Thanks! :grin:

I will merge this later unless someone says otherwise.

nicoddemus commented:
Thanks, sorry for the delay! :grin:
